File: DataSets/DataSet-New/LeetCodeTasks - Develop/LeetCodeTasks/3017_count-the-number-of-houses-at-a-certain-distance-ii/solutions/c/solution.c
#include <stdlib.h>

long long* countOfPairs(int n, int x, int y, int* returnSize) {
    *returnSize = n;
    long long* result = malloc(n * sizeof(long long));
    if (!result) return NULL;
    
    // Initialize result array to zero
    for (int i = 0; i < n; i++) {
        result[i] = 0;
    }
    
    // Ensure x <= y for simpler calculations
    if (x > y) {
        int temp = x;
        x = y;
        y = temp;
    }
    
    // Using Floyd-Warshall-like reasoning, or more simply, dynamic reachability
    int dist = y - x;
    for (int k = 1; k <= n; k++) {
        long long count = 0;
        // For each house i, calculate the distances to other houses and count the ones that are exactly k
        for (int i = 1; i <= n; i++) {
            int j1 = i + k; // Forward k
            int j2 = i - k; // Backward k
            
            // Validity check for j1
            if (j1 <= n) {
                count++;
            }
            // Validity check for j2
            if (j2 >= 1) {
                count++;
            }
        }
        
        // Also, compute using the x to y shortcut, considering how it might shortcut previous results
        // Forward and backward considering shortcut distances
        for (int i = 1; i <= n; i++) {
            // Minimum distance using x-y shortcut
            int via_x = abs(i - x) + 1 + abs(y - (i + k));
            int via_y = abs(i - y) + 1 + abs(x - (i + k));
            if (i + k <= n && (via_x == k || via_y == k)) {
                count++;
            }
            via_x = abs(i - x) + 1 + abs(y - (i - k));
            via_y = abs(i - y) + 1 + abs(x - (i - k));
            if (i - k >= 1 && (via_x == k || via_y == k)) {
                count++;
            }
        }
        
        // Record the result
        result[k - 1] = count;
    }
    
    return result;
}

File: DataSets/DataSet-New/LeetCodeTasks - Develop/LeetCodeTasks/3017_count-the-number-of-houses-at-a-certain-distance-ii/solutions/java/solution.java
class Solution {
    public long[] countOfPairs(int n, int x, int y) {
        // Ensure x is less than y for easier calculation
        if (x > y) {
            int temp = x;
            x = y;
            y = temp;
        }

        long[] result = new long[n];

        // Calculate minimum distances using the added shortcut between x and y
        for (int i = 1; i <= n; i++) {
            for (int j = i + 1; j <= n; j++) {
                int directDist = j - i;
                int viaXYDist = Math.abs(i - x) + 1 + Math.abs(y - j); // 1 for the edge between x and y
                int minDist = Math.min(directDist, viaXYDist);
                
                if (minDist < n) {
                    result[minDist]++;
                }
            }
        }

        // Each pair (i, j) is bidirectional, so multiply the count by 2
        for (int k = 1; k < n; k++) {
            result[k] *= 2;
        }

        return result;
    }
}
